### PR TITLE
Move `Signer` types out of the `signature` module

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

`Signer` types cluttering the `signature` module.

#### Summary of Changes

Move them to their own modules (preserving legacy use-paths)

### CODE REVIEW

File: sdk/src/signer/presigner.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+use {
+    crate::{
+        pubkey::Pubkey,
+        signature::Signature,
+        signer::{Signer, SignerError},
+    },
+    thiserror::Error,
+};
+#[derive(Clone, Debug, Default)]
+pub struct Presigner {
```
Comment (CriesofCarrots) [inline sdk/src/signer/presigner.rs:10]: Do you mind adding a doc comment explaining this struct? (similar to NullSigner)

File: sdk/src/signer/keypair.rs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    fn test_keypair_from_seed() {
+        let good_seed = vec![0; 32];
+        assert!(keypair_from_seed(&good_seed).is_ok());
+
+        let too_short_seed = vec![0; 31];
+        assert!(keypair_from_seed(&too_short_seed).is_err());
+    }
+
+    #[test]
+    fn test_signature_fromstr() {
```
Comment (CriesofCarrots) [inline sdk/src/signer/keypair.rs:289]: Looks like this test should stay in signature.rs ...?

**Review by CriesofCarrots — COMMENTED** [review]
I like it!
Let's use this opportunity to add module- and struct-level doc comments? In particular, I think we should document: `Signer` trait, `Presigner` struct, `Keypair` struct, and prob `keypair_from_seed()`. I can follow with some or all of this, if you like.

File: sdk/src/signer/keypair.rs (hunk 2 above)
Comment (t-nelson) [reply sdk/src/signer/keypair.rs:289]: Good eye!  I'm surprised I didn't miss more than the one

**Comment by t-nelson** [comment]
> I like it!
> Let's use this opportunity to add module- and struct-level doc comments? In particular, I think we should document: `Signer` trait, `Presigner` struct, `Keypair` struct, and prob `keypair_from_seed()`. I can follow with some or all of this, if you like.

Sure! Seems like as good a time as any

**Review by CriesofCarrots — DISMISSED** [review]
Thanks for adding all the documentation! r+ nits and feature `full`

File: sdk/src/signer/mod.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+/// specified in `Transaction` signing interfaces
+pub trait Signer {
+    /// Infallible getter of the implementor's public key. Returns the all-zeros
+    /// `Pubkey` if the implementor has none.
+    fn pubkey(&self) -> Pubkey {
+        self.try_pubkey().unwrap_or_default()
+    }
+    /// Fallible getter of the implementor's public key
+    fn try_pubkey(&self) -> Result<Pubkey, SignerError>;
+    /// Invallibly produce an Ed25519 signature over the provided `message`
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/signer/mod.rs:62]: Got the `s/v/f` too :wink:

File: sdk/src/signer/keypair.rs [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+        path::Path,
+    },
+};
+
+/// A vanilla Ed25519 key pair
+#[derive(Debug)]
+pub struct Keypair(ed25519_dalek::Keypair);
+
+impl Keypair {
+    /// Construct a new, random `Keypair` using a caller-proveded RNG
```
[the remark this answers is not in the document]
Comment (t-nelson) [reply sdk/src/signer/keypair.rs]: Damn.  I randomly chose the wrong form :slightly_smiling_face:

File: sdk/src/signer/mod.rs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+/// `Transaction` signing interfaces
+pub trait Signer {
+    /// Infallibly gets the implementor's public key. Returns the all-zeros
+    /// `Pubkey` if the implementor has none.
+    fn pubkey(&self) -> Pubkey {
+        self.try_pubkey().unwrap_or_default()
+    }
+    /// Fallibly gets the implementor's public key
+    fn try_pubkey(&self) -> Result<Pubkey, SignerError>;
+    /// Invallibly produces an Ed25519 signature over the provided `message`
```
Comment (CriesofCarrots) [inline sdk/src/signer/mod.rs:64]: Hmm, I still see `Invallibly` here... crossed commits, maybe?

Comment (CriesofCarrots) [reply sdk/src/signer/mod.rs:64]: Fix in https://github.com/solana-labs/solana/pull/17180

**Pull request merged**